File: content/5-memory/2-history-compaction/mastra/1-sliding_window.ts
import { Agent } from "@mastra/core/agent";
import { Memory } from "@mastra/memory";

const LLM_MODEL = "openai/gpt-5.4";

// sliding window: keep last N messages, discard everything older
// built into Memory — one config option, no custom code needed
const agent = new Agent({
  name: "assistant",
  instructions: "You are a helpful assistant.",
  model: LLM_MODEL,
  memory: new Memory({
    options: { lastMessages: 10 },
  }),
});

const memoryConfig = { memory: { thread: "chat_1", resource: "user_1" } };

const result1 = await agent.generate(
  "What is the capital of France?",
  memoryConfig,
);
console.log(result1.text);

const result2 = await agent.generate(
  "What is its population?",
  memoryConfig,
);
console.log(result2.text);

File: content/5-memory/2-history-compaction/pydantic/1-sliding_window.py
from pydantic_ai import Agent
from pydantic_ai.messages import ModelMessage

LLM_MODEL = "openai:gpt-5.4"

# sliding window: keep last N messages, discard everything older
# processor runs before each model call, replaces history for that call
WINDOW = 10

def sliding_window(messages: list[ModelMessage]) -> list[ModelMessage]:
    return messages[-WINDOW:]

agent = Agent(LLM_MODEL, history_processors=[sliding_window])

history: list[ModelMessage] = []

result = agent.run_sync("What is the capital of France?", message_history=history)
history = result.new_messages()
print(result.output)

result = agent.run_sync("What is its population?", message_history=history)
history = result.new_messages()
print(result.output)
